### PR TITLE
Optimize secp256r1 202306

### PR DESCRIPTION
**Pull request by contrun**
With these changes, the verification cycles for secp256r1 is now 7.8M.

### CODE REVIEW

File: src/nn/nn.c [hunk 1]
```diff
@@ -161,6 +162,9 @@ void nn_set_wlen(nn_t A, u8 new_wlen)
 	u8 i;
 	if (A->wlen == new_wlen) {
 		return;
+	} else if (A->wlen < new_wlen) {
+		A->wlen = new_wlen;
+		return;
```
Comment (XuJiandong) [inline src/nn/nn.c]: Try removing this and test the performance changes. This change might be dangerous.

File: src/nn/nn_logical.c [hunk 2]
```diff
@@ -395,19 +395,25 @@ void nn_not(nn_t A, nn_src_t B)
 	}
 }
 
-/* Count leading zeros of a word. This is constant time */
+/* Count leading zeros of a word. This is NOT constant time */
 static u8 wclz(word_t A)
```
Comment (XuJiandong) [inline src/nn/nn_logical.c:399]: There is a B extension instruction for this.

File: src/words/words_64.h [hunk 3]
```diff
@@ -25,6 +25,7 @@
 typedef uint64_t word_t;
 typedef uint32_t hword_t;
 #ifdef __GNUC__
+#define local_clz __builtin_clzl
```
Comment (XuJiandong) [inline src/words/words_64.h]: See above comment.

**Review by XuJiandong — CHANGES_REQUESTED** [review]

File: src/words/words_64.h (hunk 3 above)
Comment (mohanson) [reply src/words/words_64.h]:
```
static inline int64_t _rv64_clz(int64_t rs1) {
    int64_t rd;
    __asm__ ("clz %0, %1" : "=r"(rd) : "r"(rs1));
    return rd;
}
```

File: src/nn/nn.c (hunk 1 above)
Comment (contrun) [reply src/nn/nn.c]: The cycles consumed before reverting commit https://github.com/cryptape/libecc/pull/1/commits/b6538a012a80b9c9b1acc302a4950df42367a527
```
ckb-debugger --pprof=build/secp256r1_bench.pprof --bin build/secp256r1_bench
start main...
signature verification succeeded

Run result: 0
Total cycles consumed: 7509738(7.2M)
Transfer cycles: 15706(15.3K), running cycles: 7494032(7.1M)
```

The cycles consumed after reverting commit https://github.com/cryptape/libecc/pull/1/commits/b6538a012a80b9c9b1acc302a4950df42367a527
```
ckb-debugger --pprof=build/secp256r1_bench.pprof --bin build/secp256r1_bench
start main...
signature verification succeeded

Run result: 0
Total cycles consumed: 7569149(7.2M)
Transfer cycles: 15702(15.3K), running cycles: 7553447(7.2M)
```

**Comment by XuJiandong** [comment]
```
/riscv/lib/gcc/riscv64-unknown-linux-gnu/9.2.0/../../../../riscv64-unknown-linux-gnu/bin/ld: deps/libecc-riscv-optimized/build/libec.a(nn_logical.o): in function `nn_clz':
/code/deps/libecc-riscv-optimized/src/nn/nn_logical.c:427: undefined reference to `__clzdi2'
/riscv/lib/gcc/riscv64-unknown-linux-gnu/9.2.0/../../../../riscv64-unknown-linux-gnu/bin/ld: deps/libecc-riscv-optimized/build/libec.a(nn_logical.o): in function `nn_bitlen':
/code/deps/libecc-riscv-optimized/src/nn/nn_logical.c:447: undefined reference to `__clzdi2'
```
With old compiler (nervos/ckb-riscv-gnu-toolchain@sha256:aae8a3f79705f67d505d1f1d5ddc694a4fd537ed1c7e9622420a470d59ba2ec3), it doesn't compile.

**Comment by XuJiandong** [comment]
Suggest using this version:
```C
uint64_t clz(uint64_t n)
{
    if (n == 0) return 64;
    int c = 0;
    if (n <= 0x00000000FFFFFFFF) { c += 32; n <<= 32; };
    if (n <= 0x0000FFFFFFFFFFFF) { c += 16; n <<= 16; };
    if (n <= 0x00FFFFFFFFFFFFFF) { c += 8; n <<= 8; };
    if (n <= 0x0FFFFFFFFFFFFFFF) { c += 4; n <<= 4; };
    if (n <= 0x3FFFFFFFFFFFFFFF) { c += 2; n <<= 2; };
    if (n <= 0x7FFFFFFFFFFFFFFF) { c += 1; };
    return c;
}
```

**Pull request merged**